### PR TITLE
[image-picker] Add `videoMaxDuration` option to `launchCameraAsync`

### LINKED ISSUE #9482: ImagePicker.launchCameraAsync imposes a 10-minute limitation on video duration on iOS
### Summary of Issue 

When using ImagePicker.launchCameraAsync to record a video on iOS, the video taken is capped at a maximum duration of 10 minutes. I suspect this is due to iOS' "videoMaximumDuration" flag defaulting to 10 minutes (https://developer.apple.com/documentation/uikit/uiimagepickercontroller/1619146-videomaximumduration), but I don't see a way to override it.

### Environment - output of `expo diagnostics` & the platform(s) you're targeting

Expo CLI 3.22.1 environment info:
    System:
      OS: Windows 10 10.0.18362
    Binaries:
      Node: 12.18.0 - C:\Program Files\nodejs\node.EXE
      npm: 6.7.0 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version  3.6.0.0 AI-192.7142.36.36.6392135
    npmPackages:
      expo: ^38.0.0 => 38.0.8
      react: 16.11.0 => 16.11.0
      react-native: https://github.com/expo/react-native/archive/sdk-38.0.2.tar.gz => 0.62.2
      react-navigation: ^3.11.0 => 3.13.0

### Reproducible Demo

The repro is very simple and happens with any call to ImagePicker.launchCameraAsync that allows videos.
 
### Steps to Reproduce

1. On iOS, launch the camera using ImagePicker.launchCameraAsync
2. Start recording video
3. Wait until the video has been recording for 10 minutes

### Expected Behavior vs Actual Behavior

Expected:

There is no maximum video duration, or it is configurable through ImagePicker API

Actual:

iOS displays a message about hitting the maximum duration of 10 minutes

Is there a way to configure this limit, or can you add one?

### PR DESCRIPTION
**Pull request by barthap**
# Why

- Fixes #9482 

- Removes inconsistencies between platforms: on iOS there was default limit of 10 minutes and Android there was no limit. On Web it is unfortunately browser dependant (mostly no limit) and we have no power here.

> Interesting note. Safari was built with the same default limit as in #9482 - see [this post](https://developer.apple.com/forums/thread/655193).

- Fixes Android crash of new Kotlin `ImagePicker` on older Android versions (uncaught exception).

# How

 - Added `videoMaxDuration` option to `launchCameraAsync()` - given in seconds. Defaults to 0 - no limit.

 - Surrounded throwable call with a `try-catch` block which rejects promise.

# Test Plan

Tested on real phones: iOS 13, Android 6, Android 10

### CODE REVIEW

File: packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt [hunk 1]
```diff
@@ -266,10 +271,16 @@ class ImagePickerModule(private val mContext: Context,
       return
     }
 
-    val metadataRetriever = MediaMetadataRetriever().apply {
-      setDataSource(mContext, uri)
+    try {
+      val metadataRetriever = MediaMetadataRetriever().apply {
+        setDataSource(mContext, uri)
+      }
+      VideoResultTask(promise, uri, contentResolver, CacheFileProvider(mContext.cacheDir, ".mp4"), metadataRetriever).execute()
```
Comment (tsapeta) [inline packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:278]: @lukmccall should this always be `.mp4`? 🤔

**Review by tsapeta — APPROVED** [review]

File: docs/pages/versions/unversioned/sdk/imagepicker.md [hunk 2]
```diff
@@ -173,6 +173,7 @@ Display the system UI for taking a photo with the camera. Requires `Permissions.
 
   Option for videos:
 
+  - **videoMaxDuration (_number_)** -- Maximum duration, in seconds, for video recording. Setting this to 0 disables the limit. Defaults to 0 (no limit). **On Android**, effect of this option depends on support of installed camera app. **On Web** this option has no effect - the limit is browser-dependant.
```
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/imagepicker.md:176]: Well, you have to convince me more about this `0` default value that should be recognized as `no-limit` option.
I'm a bit skeptical here .. 🧐

File: packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m [hunk 3]
```diff
@@ -168,6 +169,8 @@ - (void)launchImagePicker:(EXImagePickerTarget)target
     }
 
     self.picker.mediaTypes = [self convertMediaTypes:self.options[@"mediaTypes"]];
+
+    self.picker.videoMaximumDuration = [[self.options objectForKey:@"videoMaxDuration"] doubleValue];
```
Comment (bbarthec) [inline packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: I'm a little concerned here 🤔 
Are you sure that passing `0` (the `unlimited` here) is forcing `UIImagePickerController` to record with no limit?
https://developer.apple.com/documentation/uikit/uiimagepickercontroller/1619146-videomaximumduration
Moreover I see that `allowsEditing` has to be `YES` in order to this `videoMaximumDuration` to take any effect (https://stackoverflow.com/questions/17375249/uiimagepickercontroller-not-respecting-video-maximum-duration-on-ipad/17375309) 🤔

**Review by bbarthec — CHANGES_REQUESTED** [review]
I'd like you to convince me about this `0` default value that should be recognized as `no-limit`. Are there any examples from around the world that propose such solution? 🤔

File: docs/pages/versions/unversioned/sdk/imagepicker.md (hunk 2 above)
Comment (tsapeta) [reply docs/pages/versions/unversioned/sdk/imagepicker.md:176]: Maybe better not to mention about `0` so people don't try to use it explicitly and only treat it as no limit (for null or undefined `doubleValue` in Obj-C would return `0.0` either way)

Comment (bbarthec) [reply docs/pages/versions/unversioned/sdk/imagepicker.md:176]: Not specifying any value on iOS would result in 10-minutes limited recording: https://developer.apple.com/documentation/uikit/uiimagepickercontroller/1619146-videomaximumduration 😅 
Dunno about Android 😉

Comment (barthap) [reply docs/pages/versions/unversioned/sdk/imagepicker.md:176]: I have reproduced current behavior - on iOS not specifying `UIImagePickerController.videoMaximumDuration` defaults it to 600 (this is the exact reason for #9482).

On Android, however, there's no standardised default limit unless you set it explicitly as intent extra. How the limit value is handled and what is its default - depends on selected camera app. Android default cam app handles it as expected - when I put the limit to intent - it stops recording, and without that intent, I gave up after ~15 minutes ;)

As `0` value seems a bit hacky - we can leave it undocumented value and just set platform defaults (see below).

File: packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m (hunk 3 above)
Comment (barthap) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: I did more research and testing, and I agree that setting it to 0 is a bit hacky and can give weird behavior (and one another problem 😞  ) when `allowsEditing` is true on iOS. The SO question about iPad you linked is only about if to show trimmer or not. I tested it deeper:

 - #### `alowsEditing = false`
    When limit set to 0, I can record video longer than 10 minutes and use it. No trimmer is shown.
   
  

   When limit is set to some value, video automatically stops recording after that time. No trimmer is shown.
  
 

---

- ####  `allowsEditing = true`

 Here the problems begin. When I set any limit, recording also stops and I can trim my video. However, an ugly message is shown at the bottom:



This ugly message, _of course,_ isn't easily configurable, but someone found a way [on SO](https://stackoverflow.com/questions/40317259/video-too-long-title-alert-shown-by-uiimagepickercontroller).

When limit is set to 0 (no-limit) - this message is shown each time I end recording, regardless of video duration. It is probably related to another [`UIVideoEditorController.videoMaximumDuration`](https://developer.apple.com/documentation/uikit/uivideoeditorcontroller/1622338-videomaximumduration), which is probably set to the same as in `UIImagePickerController`.

Video editor docs say:
> The system-enforced maximum duration for a video recording is 10 minutes; you can set this value to 10 minutes or less. The default value for this property is also 10 minutes.

It seems that we cannot support videos longer than 10 minutes with `allowsEditing=true` 🙄 

Any ideas, suggestions what to do with it?

File: packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt (hunk 1 above)
Comment (lukmccall) [reply packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:278]: Yes, cause we currently don't support other formats on Android.

File: packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m (hunk 3 above)
Comment (bbarthec) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: Thank you for this deep investigation 💪 That helps a lot 😊 
Let's separate what should be done into two cases (depending on the `allowsEditing` flag being set or not):

1. `allowsEditing = false` (default value)
- I believe the default behaviour should be `no-limit` on both platforms (it's easily achievable form what I understand),
- If on both platforms passing explicit `0` value gives us this `no-limit` behaviour then I have no position to disagree and propose other approach 🤷 Maybe to judge such scenario in our docs, let's add additional note: `Both iOS and Android platforms consume 0 value as no-limit option.`

2. `allowsEditing = true` 
- How hard it is to customize this error message? (If it's fairly easy maybe we can do sth about it?)
- Let's add additional note into docs for iOS regarding this 10-minutes platform enforced limit and maybe reject on iOS if the user provides both `allowsEditing=true` and `maxDuration > 600`?

I summon @byCedric to provide some other point of view 😉

Comment (byCedric) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: I fully agree. Using sensible defaults is always a good one to do, even when it's different on platforms. If this is a limitation from Apple then let's make the most out of it and document that limitation. When this limitation is really hard to overcome, we might be able to split the config for android and ios. So developers can use different settings on both platforms?

Comment (barthap) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: Ok, so IMO we can do it this way:
- `allowsEditing = false` - default to 0 on both platforms. If any problems would ever happen on iOS _(setting to 0 is a kind of hack after all)_, we can then fallback to 10 min limit on that platform.

- `allowsEditing = true` - defaults to 0 on Android and.. What about iOS?
  - We can set iOS default to 600 (max) and give a note in docs.
  - When user explicitly sets `maxDuration > 600` - we can reject, that's no problem.
  - When user sets it explicitly to 0 - we set it to 600: `if(maxDuration == 0) { maxDuration = 600; }` But should the user be warned somehow? If so, how?

_Also, as I am recently checking for corner cases for such options - negative values should be treated as no-limit or throw?_ 🤷 😉 

> How hard it is to customize this error message? (If it's fairly easy maybe we can do sth about it?)

I haven't tested it yet, but if [this answer](https://stackoverflow.com/a/54302086) works, it is just a category for `NSBundle`.

Comment (bbarthec) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: - Upon negative values, let's reject (check that on JS side)
- `allowsEditing = true, maxDuration = 0 (no-limit), platform = iOS` (under the hood `maxDuration = 600`) - document it and that should do the thing 👍 
- if I understand correctly having above in place would not trigger this nasty error message from UIVideoEditorController? 🤔 If so, problem solved ✅ If no, we can try this approach from SO 💪

Comment (barthap) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: > if I understand correctly having above in place would not trigger this nasty error message from UIVideoEditorController? 🤔 If so, problem solved ✅ If no, we can try this approach from SO 💪

I'm afraid we'll have to deal with it anyway. This message can be triggered (and this is expected behavior) when user sets the limit explicitly (e.g. to 30s) and when maximum duration is exceeded.

Comment (bbarthec) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: Ok, let's do it then, shall we? 😊

Comment (barthap) [reply packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:173]: Unless it's rejected by Apple 🙄  cc @lukmccall

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 4]
```diff
@@ -11,6 +11,26 @@ import {
...
 } from './ImagePicker.types';
 
+function validateOptions(options: ImagePickerOptions) {
+  const { aspect, quality, videoMaxDuration } = options;
+
+  if (aspect != null) {
+    const [x, y] = aspect;
+
+    if (x <= 0 || y <= 0) {
+      throw new CodedError('ERR_INVALID_ARGUMENT', 'Invalid aspect ratio values');
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:21]:
```suggestion
      throw new CodedError('ERR_INVALID_ARGUMENT', `Invalid aspect ratio values ${x}:${y}. Provide positive numbers.`);
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 5]
```diff
@@ -11,6 +11,26 @@ import {
...
+    }
+  }
+
+  if (quality && (quality < 0 || quality > 1)) {
+    throw new CodedError('ERR_INVALID_ARGUMENT', 'Quality value must be between 0 and 1');
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:26]:
```suggestion
    throw new CodedError('ERR_INVALID_ARGUMENT', `Invalid 'quality' value ${quality}. Provide a value between 0 and 1.`);
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 6]
```diff
@@ -11,6 +11,26 @@ import {
...
+  }
+
+  if (videoMaxDuration && videoMaxDuration < 0) {
+    throw new CodedError('ERR_INVALID_ARGUMENT', 'videoMaxDuration must be a non-negative number');
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:30]:
```suggestion
    throw new CodedError('ERR_INVALID_ARGUMENT', `Invalid 'videoMaxDuration' value ${videoMaxDuration}. Provide a non-negative number.`);
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 7]
```diff
@@ -11,6 +11,26 @@ import {
...
+  }
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:31]:
```suggestion
  }
  return { aspect, quality, videoMaxDuration };
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 8]
```diff
@@ -42,6 +64,8 @@ export async function launchCameraAsync(
   if (!ExponentImagePicker.launchCameraAsync) {
     throw new UnavailabilityError('ImagePicker', 'launchCameraAsync');
   }
+
+  validateOptions(options);
   return await ExponentImagePicker.launchCameraAsync(options);
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:69]:
```suggestion
  return await ExponentImagePicker.launchCameraAsync(validateOptions(options));
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 9]
```diff
@@ -42,6 +64,8 @@ export async function launchCameraAsync(
   if (!ExponentImagePicker.launchCameraAsync) {
     throw new UnavailabilityError('ImagePicker', 'launchCameraAsync');
   }
+
+  validateOptions(options);
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:68]:
```suggestion
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 10]
```diff
@@ -33,6 +53,8 @@ export async function launchImageLibraryAsync(
   if (!ExponentImagePicker.launchImageLibraryAsync) {
     throw new UnavailabilityError('ImagePicker', 'launchImageLibraryAsync');
   }
+
+  validateOptions(options);
   return await ExponentImagePicker.launchImageLibraryAsync(options);
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:58]:
```suggestion
  return await ExponentImagePicker.launchImageLibraryAsync(validateOptions(options));
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 11]
```diff
@@ -33,6 +53,8 @@ export async function launchImageLibraryAsync(
   if (!ExponentImagePicker.launchImageLibraryAsync) {
     throw new UnavailabilityError('ImagePicker', 'launchImageLibraryAsync');
   }
+
+  validateOptions(options);
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:57]:
```suggestion
```

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 12]
```diff
@@ -11,6 +11,26 @@ import {
   VideoExportPreset,
 } from './ImagePicker.types';
 
+function validateOptions(options: ImagePickerOptions) {
+  const { aspect, quality, videoMaxDuration } = options;
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:15]:
```suggestion
function validateOptions({ aspect, quality, videoMaxDuration }: ImagePickerOptions) {
```

File: packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m [hunk 13]
```diff
@@ -173,9 +174,24 @@ - (void)launchImagePicker:(EXImagePickerTarget)target
       self.picker.videoExportPreset = [self importVideoExportPreset:self.options[@"videoExportPreset"]];
     }
     
+    NSTimeInterval videoMaxDuration = [[self.options objectForKey:@"videoMaxDuration"] doubleValue];
+    
     if ([[self.options objectForKey:@"allowsEditing"] boolValue]) {
       self.picker.allowsEditing = true;
+      
+      if (videoMaxDuration > 600.0) {
+        self.reject(@"ERR_IMAGE_PICKER_MAX_DURATION", @"`videoMaxDuration` limits to 600 when `allowsEditing=true`", nil);
```
Comment (bbarthec) [inline packages/expo-image-picker/ios/EXImagePicker/EXImagePicker.m:183]:
```suggestion
        self.reject(@"ERR_IMAGE_PICKER_MAX_DURATION", @"'videoMaxDuration' limits to 600 when 'allowsEditing=true'", nil);
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
Ensure that you do not change `react-native-lab/react-native` as this is not a part of this PR, right? 🤔
Some suggestions about error messages: provide what's wrong in simplest manner available and if possible solution to the error in simple sentence. Examples in suggestions.

File: packages/expo-image-picker/src/ImagePicker.ts (hunk 7 above)
Comment (barthap) [reply packages/expo-image-picker/src/ImagePicker.ts:31]: I need to return whole `options` object, not only three params.

File: packages/expo-image-picker/src/ImagePicker.ts (hunk 12 above)
Comment (barthap) [reply packages/expo-image-picker/src/ImagePicker.ts:15]: If I do it this way, I won't be able to return `options` object, and thus inline `validateOptions()` as you suggested

**Review by bbarthec — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]

**Comment by robsonbbs** [comment]
I know that's very old but let me add my 2 cents: 

This option's a little bit confusing since `ImagePickerOptions` is present on `launchImageLibraryAsync` as well. 
I was led to believe that setting this with `launchImageLibraryAsync` would result in filtering videos by the duration that I had set. I would consider having 2 different types.

**Pull request merged**